internal/security/bootstrapper/command/setupacl/command.go

```diff
@@ -0,0 +1,364 @@
+		return interfaces.StatusCodeExitWithError, fmt.Errorf("failed to configure Consul access: %v", err)
+	}
+
+	return
+}
+
+// GetCommandName returns the name of this command
+func (c *cmd) GetCommandName() string {
+	return CommandName
+}
+
+func (c *cmd) getAPI_URL(api string) (string, error) {
+	apiURL := fmt.Sprintf("%s://%s:%d%s", c.configuration.StageGate.Registry.ACL.Protocol,
+		c.configuration.StageGate.Registry.Host, c.configuration.StageGate.Registry.Port, api)
+	_, err := url.Parse(apiURL)
+	if err != nil {
+		return "", fmt.Errorf("failed to parse API URL: %v", err)
+	}
+	return apiURL, nil
+}
+
+// waitForNonEmptyConsulLeader is a special waitFor function on waiting for "non-empty" leader being available
+// the ordinary http waitFor won't work as the returned http status code from API call is 200 even when Consul's leader
+// is an empty string ("") but we need an non-empty leader; so 200 doesn't mean we have a leader
+func (c *cmd) waitForNonEmptyConsulLeader() error {
+	c.errs = make(chan error, 1)
+	// check Consul whether we have non-empty leader elected before proceed to Consul's ACL bootstrapping process
+	// as it requires non-empty leader to start with
+	go func() {
+		c.checkConsulLeader()
+		c.waitGroup.Wait()
+	}()
+
+	gotLeader := make(chan bool)
+	// a separate goroutine to keep waiting for a leader got elected from the checkConsulLeader goroutine
+	// until retryTimeout is reached otherwise
+	go func() {
+		for {
+			if err := <-c.errs; err != nil {
+				continue
+			}
+			gotLeader <- true
+			close(gotLeader)
+			return
+		}
+	}()
+
+	// block here until either we get a leader, or timeout is reached in which case ACL bootstrapping cannot continue
+	select {
+	case <-gotLeader:
+		c.loggingClient.Info("found Consul leader to bootstrap ACL")
+		return nil
+	case <-time.After(c.retryTimeout):
+		return errors.New("timed out to get non-empty Consul leader")
+	}
+}
+
+// checkConsulLeader is a goroutine constantly running if there is an error from API call
+func (c *cmd) checkConsulLeader() {
+	c.waitGroup.Add(1)
+	go func() {
+		defer c.waitGroup.Done()
+		ticker := time.NewTicker(time.Second)
+		defer ticker.Stop()
+		// ticks every second to retry if we receive an error from API call
+		for range ticker.C {
+			if err := c.callConsulLeaderAPI(); err != nil {
+				c.loggingClient.Warnf("error from getting Consul leader API call, will retry it again: %v", err)
+				c.errs <- err
+				continue
+			}
+			// once reached here, we have a leader and terminate this goroutine
+			c.errs <- nil
+			return
+		}
+	}()
+}
+
+// callConsulLeaderAPI makes http request call to get the registry Consul leader
+// the response of getting leader call could be an empty leader (represented by "")
+// even if the http status code is 200 when Consul is just booting up and
+// it will take a bit of time to elect the raft leader
+func (c *cmd) callConsulLeaderAPI() error {
```

Comment:
how about `createConsulLeader()` for the name?

Reply:
we not "creating" though... this funciton is actually calling consul's get leader api

Reply:
The result of call the API is the leader is create or set... callXyzApi is not a descriptive name of have the real action is.

Reply:
ok